File: Java/Sudoku/src/Sudoku.java
/**
 * @author Twenty Codes, LLC
 * @author ricky barrette
 * @date Dec 7, 2010
 */

/**
 * This will be a sudoku puzzle
 * @author ricky barrette
 */
public class Sudoku {
	
	private int[][] mPuzzle;
	private final boolean DEBUG = true;

	/**
	 * Creates a new Sudoku
	 * @author ricky barrette
	 * @param puzzle 
	 */
	public Sudoku(int[][] puzzle) {
		mPuzzle = puzzle;
	}
	
	/**
	 * creates a copy of the maze that we can modify and discard
	 * @return a copy of the maze
	 */
	private int[][] makeTempMaze() {
		int[][] m = new int[mPuzzle.length][mPuzzle[0].length];
		for (int i = 0; i < m.length; i++) {
			for (int index = 0; index < m[0].length; index++) {
				m[i][index] = mPuzzle[i][index];
			}
		}
		return m;
	}
	
	/**
	 * solves the puzzle
	 * @return true if the puzzle is solvable
	 * @author ricky barrette
	 */
	public boolean solve(){
		return solve(0,0, mPuzzle);
	}
	
	/**
	 * Recursively solves a sudoku puzzle 
	 * @return true if the puzzle is solvable
	 * @author ricky barrette
	 */
	private boolean solve(int row, int column, int[][] temp){
		
		/*
		 * move on to the next row, if we have no more columns
		 */
		if (column == 9) {
            column = 0;
            if (++row == 9)
                return true;
        }
		
		/*
		 * if this spot is filled in, then lets move on
		 */
		if (temp[row][column] != 0)
			solve(row, column +1, temp);
		
		/*
		 * we are going to try every number, to find one that is valid.
		 */
		for(int i = 1; i <= 9; ++i){
			if(isValid(i, row, column, temp)){
				temp[row][column] =  i;
				if(solve(row, column +1, temp))
					return true;
			}
		}
		
		/*
		 * reset on backtrack
		 */
		temp[row][column] = 0;
	
		/*
		 * XXX This is only for testing comment out for production
		 * this allows us to watch the maze being solved
		 */
		if (DEBUG) {
			try {
				Thread.sleep(50);
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
			System.out.println(puzzleToString(temp));
		}
		return false;
	}
	
	/**
	 * checks to see if the value is a valid answer for this location
	 * @param newValue in question
	 * @param row of newValue
	 * @param column of newValue
	 * @param temp puzzle
	 * @return true if the newValue is a valid answer
	 * @author ricky barrette
	 */
	private boolean isValid(int newValue, int row, int column, int[][] temp) {
		
		/*
		 * row
		 */
		for (int i = 0; i < 9; ++i)
            if (newValue == temp[i][column])
                return false;

		/*
		 * column
		 */
        for (int i = 0; i < 9; ++i)
            if (newValue == temp[row][i])
                return false;

        int boxRowOffset = (row / 3)*3;
        int boxColOffset = (column / 3)*3;
        
        /*
         * box
         */
        for (int i = 0; i < 3; ++i)
            for (int index = 0; index < 3; ++index)
                if (newValue == temp[boxRowOffset+i][boxColOffset+index])
                    return false;

        /*
         * no violations, so it's legal
         */
        return true;
	}

	/**
	 * converts the maze into a human readable string
	 * @param maze
	 * @return string for of the maze
	 */
	private String puzzleToString(int[][] puzzle){
		StringBuffer s = new StringBuffer();
		s.append("\n");
		for (int i = 0; i < puzzle.length; i++) {
			for (int index = 0; index < puzzle[0].length; index++) {
				s.append(puzzle[i][index] + " ");
			}
			s.append("\n");
		}
		return s.toString();
	}
	
	/**
	 * returns a human readable String of this maze
	 * (non-Javadoc)
	 * @see java.lang.Object#toString()
	 */
	@Override
	public String toString(){
		return puzzleToString(mPuzzle);
	}

}

File: C/Sudoku/sudoku.cpp
 
/**
 * sudoku.cpp
 * 
 * This program will be used to solve sudoku puzzles
 * 
 * @date oct, 25, 2011
 * @author ricky barrette
 */

#include <iostream>
#include "sudoku.h"

int mPuzzle[PUZZLE_SIZE][PUZZLE_SIZE] = {	{6,0,0, 8,3,0, 0,4,0},
						{0,8,5, 1,0,0, 0,0,0},
						{0,0,0, 0,0,0, 0,0,0},
		    
						{4,0,0, 0,6,0, 0,8,0},
						{0,0,0, 0,0,8, 0,0,3},
						{1,6,0, 0,0,2, 0,0,0},
					    
						{0,0,9, 0,0,3, 0,1,0},
						{0,5,0, 6,0,0, 0,0,0},
						{0,1,0, 0,4,0, 2,9,0}	};

int mCount = 0;

/**
 * Main function
 */
int main(){
  std::cout << "\nHello World, Sudoku!\n\n\n";
  
  printPuzzle(mPuzzle);
  
  if(solve(mPuzzle) == 0) {
    std::cout << "NO SOLUTION\n";
    std::cout << mCount;
   std::cout << " trys\n\n";
  } else {
   std::cout << "\nSolved in ";
   std::cout << mCount;
   std::cout << " trys\n\n";
   
   printPuzzle(mPuzzle);
  }
  return 1;
}

/**
 * Checks the Row
 * @returns 1 if legal, 0 if illegal
 * @author ricky barrette
 */
int checkRow(int n, int r, int puzzle[PUZZLE_SIZE][PUZZLE_SIZE]){
  for(int index = 0; index < PUZZLE_SIZE; index++)
      if(n == puzzle[r][index])
	return 0;
   return 1;
}

/**
 * Checks the  Column
 * @returns 1 if legal, 0 if illegal
 * @author ricky barrette
 */
int checkColumn(int n, int c, int puzzle[PUZZLE_SIZE][PUZZLE_SIZE]){
  for(int index = 0; index < PUZZLE_SIZE; index++)
      if(n == puzzle[index][c])
	return 0;
   return 1;
}

/**
 * Checks the "Box"
 * @returns 1 if legal, 0 if illegal
 * @author ricky barrette
 */
int checkBox(int n, int r, int c, int puzzle[PUZZLE_SIZE][PUZZLE_SIZE]){
  r = (r/BOX_SIZE)*BOX_SIZE;
  c = (c/BOX_SIZE)*BOX_SIZE;
  
  for(int ri = 0; ri < BOX_SIZE; ri++)
    for(int ci = 0; ci < BOX_SIZE; ci++)
	if(n == puzzle[r+ri][c+ci])
	  return 0;
  
  //no violations
  return 1;
}

/**
 * Solves a sudoku puzzle
 * @returns 1 if solvable, 0 if not
 * @author ricky barrette
 */
int solve(int puzzle[PUZZLE_SIZE][PUZZLE_SIZE]){
  return solve(0,0, puzzle);
}

/**
 * Solves a sudoku puzzle
 * @returns 1 if solvable, 0 if not
 * @author ricky barrette
 */
int solve(int r, int c, int puzzle[PUZZLE_SIZE][PUZZLE_SIZE]){
  
  if(DEBUG){
    sleep(1);
    printPuzzle(puzzle);
  }
  
  mCount++;
   
  if (r == PUZZLE_SIZE) {
    r = 0;
    if (++c == PUZZLE_SIZE)
      return 1;
  }
  //skip filled cells
  if (puzzle[r][c] != 0)
    return solve(r+1,c,puzzle);

  for (int n = 1; n <= PUZZLE_SIZE; n++) {
    if (legal(n,r,c,puzzle) == 1) {
      puzzle[r][c] = n;
      if (solve(r+1,c,puzzle) == 1)
	return 1;
    }
  }
  //reset on backtrack
  puzzle[r][c] = 0;
  return 0;
}

/**
 * Prints the puzzle onto the screen
 * @author ricky barrette
 */
void printPuzzle(int puzzle[PUZZLE_SIZE][PUZZLE_SIZE]){
  
  for(int ri = 0; ri < PUZZLE_SIZE; ri++){
    for(int ci = 0; ci < PUZZLE_SIZE; ci++){ 
      std::cout << puzzle[ri][ci];
      //format: break for groups of BOX_SIZE
      if(ci % BOX_SIZE == 2)
	std::cout <<"  ";
    }
    //format: break columns of BOX_SIZE
    if(ri % BOX_SIZE == 2)
	std::cout <<"\n";
    //format: break for new row
    std::cout << "\n";
  }
  std::cout <<"\n-------------\n";
}

/**
 * Checks the puzzle if it is legal
 * @returns 1 if legal; 0 if not
 * @author ricky barrette
 */
int legal(int n, int r, int c, int puzzle[PUZZLE_SIZE][PUZZLE_SIZE]){
  if(checkRow(n, r, puzzle) == 1)
    if(checkColumn(n, c, puzzle) == 1)
      if(checkBox(n, r, c, puzzle) == 1)
	return 1;
  return 0;
}
